src/draw/lv_draw_gradient.c

```diff
@@ -0,0 +1,316 @@
+/**
+ * @file lv_draw_gradient.c
+ *
+ */
+
+/*********************
+ *      INCLUDES
+ *********************/
+#include "lv_draw_gradient.h"
+
+
+/*********************
+ *      DEFINES
+ *********************/
+#if DITHER_GRADIENT
+    #define GRAD_CM(r,g,b) LV_COLOR_MAKE32(r,g,b)
+    #define GRAD_CONV(t, x) t.full = lv_color_to32(x)
+#else
+    #define GRAD_CM(r,g,b) LV_COLOR_MAKE(r,g,b)
+    #define GRAD_CONV(t, x) t = x
+#endif
+
+#define MAX_WIN_RES     1024 /**TODO: Find a way to get this information: max(horz_res, vert_res)*/
+
+#if DITHER_GRADIENT
+    #if LV_DITHER_ERROR_DIFFUSION == 1
+        #define LV_DEFAULT_GRAD_CACHE_SIZE  sizeof(lv_gradient_cache_t) + MAX_WIN_RES * sizeof(lv_grad_color_t) + MAX_WIN_RES * sizeof(lv_color_t) + MAX_WIN_RES * sizeof(lv_scolor24_t)
+    #else
+        #define LV_DEFAULT_GRAD_CACHE_SIZE  sizeof(lv_gradient_cache_t) + MAX_WIN_RES * sizeof(lv_grad_color_t) + MAX_WIN_RES * sizeof(lv_color_t)
+    #endif /* LV_DITHER_ERROR_DIFFUSION */
+#else
+    #define LV_DEFAULT_GRAD_CACHE_SIZE  sizeof(lv_gradient_cache_t) + MAX_WIN_RES * sizeof(lv_grad_color_t)
+#endif /* DITHER_GRADIENT */
+
+/**********************
+ *  STATIC PROTOTYPES
+ **********************/
+static lv_gradient_cache_t * next_in_cache(lv_gradient_cache_t * first);
+
+typedef lv_res_t (*op_cache_t)(lv_gradient_cache_t * c, void * ctx);
+static lv_res_t iterate_cache(op_cache_t func, void * ctx, lv_gradient_cache_t ** out);
+static size_t get_cache_item_size(lv_gradient_cache_t * c);
+static lv_gradient_cache_t * allocate_item(const lv_gradient_t * g, lv_coord_t w, lv_coord_t h);
+static lv_res_t find_oldest_item_life(lv_gradient_cache_t * c, void * ctx);
+static lv_res_t kill_oldest_item(lv_gradient_cache_t * c, void * ctx);
+static lv_res_t find_item(lv_gradient_cache_t * c, void * ctx);
+static void free_item(lv_gradient_cache_t * c);
+static  uint32_t compute_key(const lv_gradient_t * g, lv_coord_t w, lv_coord_t h);
+
+
+/**********************
+ *   STATIC VARIABLE
+ **********************/
+static uint8_t * grad_cache_mem = 0;
+static size_t    grad_cache_size = 0;
+static uint8_t * grad_cache_end = 0;
+
+/**********************
+ *   STATIC FUNCTIONS
+ **********************/
+union void_cast {
+    const void * ptr;
+    const uint32_t value;
+};
+
+uint32_t compute_key(const lv_gradient_t * g, lv_coord_t size, lv_coord_t w)
```

Comment:
These functions are in the `STATIC FUNCTIONS` section. So isn't there missing `static` keywords here?

Reply:
Well, don't want to start a debate here, but you don't need to repeat `static` if the prototype contains the `static` intenal linkage. 

## C11: 6.2.2 Linkages of identifiers (p4):

    For an identifier declared with the storage-class specifier extern in a scope in which a prior declaration of that identifier is visible, **if the prior declaration specifies internal or external linkage, the linkage of the identifier at the later declaration is the same as the linkage specified at the prior declaration**. [...]

But to be consistent with the rest of the code, I'll add static here too.

Reply:
I didn't know that, thanks. 

> But to be consistent with the rest of the code, I'll add static here to

Thanks.